presto-benchmark-runner/src/main/java/com/facebook/presto/benchmark/framework/BenchmarkManager.java

```diff
@@ -0,0 +1,173 @@
+        this.maxConcurrency = config.getMaxConcurrency();
+        this.benchmarkResubmissionLimit = config.getBenchmarkResubmissionLimit();
+        this.eventClients = ImmutableSet.copyOf(eventClients);
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        this.executor = newFixedThreadPool(maxConcurrency);
+        this.completionService = new ExecutorCompletionService<>(executor);
+        BenchmarkSuite benchmarkSuite = benchmarkSuiteSupplier.get();
+        Map<String, BenchmarkQuery> queries = benchmarkSuite.getQueryMap();
+        BenchmarkSuiteInfo benchmarkSuiteInfo = benchmarkSuite.getSuiteInfo();
+        List<PhaseSpecification> phases = benchmarkSuiteInfo.getPhases();
+        for (PhaseSpecification phase : phases) {
+            if (phase.getExecutionStrategy() == CONCURRENT) {
+                submit(ImmutableList.copyOf(queries.values()), benchmarkSuiteInfo.getSessionProperties());
+            }
+        }
+        reportProgressUntilFinished();
+    }
+
+    @VisibleForTesting
+    AtomicInteger getQueriesSubmitted()
+    {
+        return queriesSubmitted;
+    }
+
+    @Override
+    public synchronized boolean resubmit(BenchmarkQueryRunner benchmarkQueryRunner, BenchmarkQueryException benchmarkQueryException)
+    {
+        if (!shouldResubmit(benchmarkQueryException)) {
+            return false;
+        }
+
+        String name = benchmarkQueryRunner.getQuery().getName();
+        Integer resubmittedCount = resubmittedCounts.getOrDefault(name, 0);
+        if (resubmittedCount >= benchmarkResubmissionLimit) {
+            log.info("Benchmark query %s failed with %s, resubmission limit exceeded", name, benchmarkQueryException.getErrorCode());
+            return false;
+        }
+
+        queriesSubmitted.addAndGet(1);
+        resubmittedCounts.compute(name, (key, count) -> count == null ? 1 : count + 1);
+        completionService.submit(benchmarkQueryRunner::run);
+        log.info("Benchmark query %s failed with %s, resubmitted (%s/%s)", name, benchmarkQueryException.getErrorCode(), resubmittedCount, benchmarkResubmissionLimit);
+        return true;
+    }
+
+    private void submit(List<BenchmarkQuery> benchmarkQueries, Map<String, String> sessionProperties)
+    {
+        for (BenchmarkQuery benchmarkQuery : benchmarkQueries) {
+            BenchmarkQueryRunner benchmarkQueryRunner = benchmarkQueryRunnerFactory.get(this, benchmarkQuery, sessionProperties);
+            completionService.submit(benchmarkQueryRunner::run);
+        }
+        log.info("Queries submitted: %s", benchmarkQueries.size());
+        this.queriesSubmitted.addAndGet(benchmarkQueries.size());
+    }
+
+    private void reportProgressUntilFinished()
+    {
+        int completed = 0;
+        double lastProgress = 0;
+        Map<EventStatus, Integer> statusCount = new EnumMap<>(EventStatus.class);
+
+        while (completed < queriesSubmitted.get()) {
```

Comment:
If any query fails, the run should fail by exiting with non-zero exit code.

Reply:
@caithagoras I don't think we should fail the whole run if any query fails. For custom workloads like the Facebook one, it's common that some queries might fail. The whole run is still valuable with most of the queries succeed. If we make the success criteria as all queries should succeed, it might be hard to get a successful run. However when reporting the data, we may want to exclude such runs with partial success.

Reply:
As discussed, we will provide a configurable parameter 'proceed-with-failures' that tells us whether to fail the entire the run or proceed with executing remaining queries/phases.